src/main/java/org/junitpioneer/jupiter/CartesianEnumSource.java

```diff
@@ -0,0 +1,172 @@
+/*
+ * Copyright 2016-2021 the original author or authors.
+ *
+ * All rights reserved. This program and the accompanying materials are
+ * made available under the terms of the Eclipse Public License v2.0 which
+ * accompanies this distribution and is available at
+ *
+ * http://www.eclipse.org/legal/epl-v20.html
+ */
+
+package org.junitpioneer.jupiter;
+
+import static java.util.Arrays.stream;
+import static java.util.stream.Collectors.toSet;
+
+import java.lang.annotation.Documented;
+import java.lang.annotation.ElementType;
+import java.lang.annotation.Repeatable;
+import java.lang.annotation.Retention;
+import java.lang.annotation.RetentionPolicy;
+import java.lang.annotation.Target;
+import java.util.Objects;
+import java.util.Set;
+import java.util.function.BiConsumer;
+import java.util.function.BiPredicate;
+import java.util.regex.Pattern;
+import java.util.regex.PatternSyntaxException;
+
+import org.junit.jupiter.params.provider.ArgumentsSource;
+import org.junit.platform.commons.PreconditionViolationException;
+
+/**
+ * {@code @CartesianEnumSource} is an argument source for constants of a
+ * specified {@linkplain #value Enum}.
+ *
+ * <p>The enum constants will be provided as an argument source to
+ * the corresponding parameter of the annotated {@code @CartesianProductTest} method.
+ *
+ * <p>The set of enum constants can be restricted via the {@link #names} and
+ * {@link #mode} attributes.
+ *
+ * <p>This annotation is {@link Repeatable}. You should declare one
+ * {@code @CartesianEnumSource} per parameter.
+ * </p>
+ *
+ * @see CartesianProductTest
+ *
+ * @since 1.2.1
+ */
+@Target({ ElementType.ANNOTATION_TYPE, ElementType.METHOD })
+@Retention(RetentionPolicy.RUNTIME)
+@Documented
+@Repeatable(CartesianEnumSource.CartesianEnumSources.class)
+@ArgumentsSource(CartesianEnumArgumentsProvider.class)
+public @interface CartesianEnumSource {
+
+	/**
+	 * The enum type that serves as the source of the enum constants.
+	 *
+	 * @see #names
+	 * @see #mode
+	 */
+	Class<? extends Enum<?>> value();
```

Comment:
Opening a thread not to leave this behind - this line causes a Sonar [code smell](https://sonarcloud.io/project/issues?id=junit-pioneer_junit-pioneer&issues=AXcIQv9wtsx-iI78REli&open=AXcIQv9wtsx-iI78REli&pullRequest=409). 

Instead of polluting the annotation with `//NOSONAR`, I suggest closing the issue on Sonar as [false positive](https://docs.sonarqube.org/latest/faq/).

Reply:
I've just marked it as _won't fix_. I'm tempted to deactivate the rule as a whole - maybe if it pops up more often.